src/daft-scan/src/scan_task_iters/mark_scan_tasks_for_split.rs

```diff
@@ -0,0 +1,113 @@
+use std::{ops::Deref, sync::Arc};
+
+use common_daft_config::DaftExecutionConfig;
+use common_error::DaftResult;
+use common_file_formats::{FileFormatConfig, ParquetSourceConfig};
+
+use super::{
+    retrieve_parquet_metadata::NeedsParquetMetadata,
+    split_parquet_files_by_rowgroup::SplittableScanTaskRef, BoxScanTaskIter,
+};
+use crate::{storage_config::StorageConfig, DataSource, ScanTaskRef};
+
+/// Decides whether or not ScanTasks need to be split
+///
+/// Returns an iterator of [`ScanTaskSplitDecision`], which implement [`NeedsParquetMetadata`]. These can then be fed into
+/// [`super::retrieve_parquet_metadata::batched_parquet_metadata_retrieval`] which will transform the [`ScanTaskSplitDecision`]
+/// into [`ScanTaskSplitByRowgroup`] by fetching the Parquet metadatas.
+pub(crate) fn decide_scantask_split_by_rowgroups<'a>(
+    scan_tasks: BoxScanTaskIter<'a>,
+    config: &'a DaftExecutionConfig,
+) -> impl Iterator<Item = DaftResult<ScanTaskSplitDecision>> + 'a {
+    scan_tasks.map(|st| Ok(ScanTaskSplitDecision::from_scan_task(st?, config)))
+}
+
+/// Wraps a ScanTaskRef, indicating whether or not it should be split by Parquet Rowgroups
+pub(crate) enum ScanTaskSplitDecision {
+    NoSplit(ScanTaskRef),
+    Split(ScanTaskRef),
+}
+
+impl ScanTaskSplitDecision {
+    /// Optionally creates [`SplitScanTaskInfo`] depending on whether or not a given ScanTask can be split
+    pub fn from_scan_task(scan_task: ScanTaskRef, config: &DaftExecutionConfig) -> Self {
+        /* Only split parquet tasks if they:
+            - have one source
+            - use native storage config
+            - have no specified chunk spec or number of rows
+            - have size past split threshold
+            - no iceberg delete files
+        */
+        if let (
+            FileFormatConfig::Parquet(ParquetSourceConfig { .. }),
+            StorageConfig::Native(_),
+            [source],
+            Some(None),
+            None,
+            est_materialized_size,
+        ) = (
+            scan_task.file_format_config.as_ref(),
+            scan_task.storage_config.as_ref(),
+            &scan_task.sources[..],
+            scan_task.sources.first().map(DataSource::get_chunk_spec),
+            scan_task.pushdowns.limit,
+            scan_task.estimate_in_memory_size_bytes(Some(config)),
+        ) && est_materialized_size.map_or(true, |est| est > config.scan_tasks_max_size_bytes)
+            && source
+                .get_iceberg_delete_files()
+                .map_or(true, std::vec::Vec::is_empty)
+        {
+            Self::Split(scan_task)
+        } else {
+            Self::NoSplit(scan_task)
+        }
+    }
+}
+
+impl Deref for ScanTaskSplitDecision {
+    type Target = ScanTaskRef;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            Self::NoSplit(st) | Self::Split(st) => st,
+        }
+    }
+}
+
+impl NeedsParquetMetadata for ScanTaskSplitDecision {
```

Comment:
Is there a reason we have this organized in a trait? I only see one implementation of this

Reply:
https://github.com/Eventual-Inc/Daft/pull/3454#discussion_r1884373893